### ISSUE TITLE
Is there any plan to implement a read-wirte lock for lua_shared_dict

### ISSUE THREAD

**Issue by lxxstc** [openresty/lua-nginx-module#656]
I was working on a dynamic upstream feature for Marathon-docker framework  with balancer_by_lua_block,  which upstream server list stored in lua_shared_dict.  At this case, server list reading must be blocked when reloading the server list, and it's not a good idea to get a lock at each http request. I think, that could be properly sovled with a read-write lock.
So, do you have any plan to implement a read-wirte lock for lua_shared_dict

**Comment by agentzh** [comment]
@lxxstc Yes, there're plans to implement the readers-writer lock. Patches welcome.

BTW, for your use case, I don't think the existing locking can become a problem. Better do realistic benchmark and measurement instead of guessing. Shm rbtree traversal is usually very fast, since it's just memory accesses, no syscalls, no context switches. Locking can only become a problem when you have huge number of keys and you have many nginx workers competing for the keys in the same shm store.
